xdsl/ir.py

```diff
@@ -421,8 +446,21 @@ def verify(self, verify_nested_ops: bool = True) -> None:
             for region in self.regions:
                 region.verify()
 
+    # TODO replace with trait
     def verify_(self) -> None:
-        pass
+        try:
+            assert self.lhs
+        except AttributeError:
+            return
+
+        if self.lhs.typ != self.rhs.typ or self.rhs.typ != self.result.typ:
```

Comment:
I don't agree with this verifier. AFAIU, we can have operations with different input and result types, like for example a `convert` operation that converts i32 to i64 or something.

I might be missing something here feel free to point out!

Reply:
Actually, the fields `lhs` and `rhs` are not even guaranteed to be there, right? And if there is a `lhs`, this is not guaranteed to be a binary operation with the same input and result types.

Reply:
> 

This happens in scf_ops.xdsl right?

Reply:
> Actually, the fields `lhs` and `rhs` are not even guaranteed to be there, right? And if there is a `lhs`, this is not guaranteed to be a binary operation with the same input and result types.

That is why I have the asserts which I consider awful

Reply:
> This happens in scf_ops.xdsl right?

Yes, there is an example of that in the `arith.cmpi` operation.

Reply:
> > Actually, the fields `lhs` and `rhs` are not even guaranteed to be there, right? And if there is a `lhs`, this is not guaranteed to be a binary operation with the same input and result types.
> 
> That is why I have the asserts which I consider awful

In the `BinaryOperation` subclass design I proposed, you won't need the assert as you can enforce the `BinaryOperation` to have the needed fields. That's another reason why I feel this design would be cleaner.

Reply:
I agree, we should definitely not put this code here, and we should use `BinaryOperation` instead.

Reply:
I like the idea of BinaryOperation, thanks!